cmake/global_functions.cmake

```diff
@@ -0,0 +1,266 @@
+##################################################################
+#
+# cFS Mission global CMake function definitions
+#
+# This is included by the top-level script and can define
+# common routines and variables that may be referenced in both 
+# the mission (non-arch) and arch-specific builds
+#
+##################################################################
+
+include(CMakeParseArguments)
+
+##################################################################
+#
+# FUNCTION: generate_c_headerfile
+#
+# Generates a C header file in the build directory.
+# First argument is the file name to write.  All remaining arguments will be
+# concatenated and written to the file.
+#
+function(generate_c_headerfile FILE_NAME)
+
+    # Determine the absolute location for this wrapper file
+    # If it is relative path then assume it is relative to ${CMAKE_BINARY_DIR}
+    # This should not write generated files to ${CMAKE_SOURCE_DIR}
+    if (NOT IS_ABSOLUTE "${FILE_NAME}")
+        set(FILE_NAME "${CMAKE_BINARY_DIR}/${FILE_NAME}")
+    endif (NOT IS_ABSOLUTE "${FILE_NAME}")
+
+    # Generate an include guard
+    get_filename_component(FILE_GUARD "${FILE_NAME}" NAME)
+    string(REGEX REPLACE "[^A-Za-z0-9]" "_" FILE_GUARD "${FILE_GUARD}")
+    string(TOUPPER "GENERATED_INCLUDE_${FILE_GUARD}" FILE_GUARD)
+    set(GENERATED_FILE_HEADER
+        "/* Generated header file.  Do not edit */\n\n"
+        "#ifndef ${FILE_GUARD}\n"
+        "#define ${FILE_GUARD}\n\n"
+    )
+    set(GENERATED_FILE_TRAILER
+        "#endif /* ${FILE_GUARD} */\n"
+    )
+
+    # Use configure_file() to write the file, as this automatically
+    # has the logic to not update the timestamp on the file unless it changes.
+    string(REPLACE ";" "" GENERATED_FILE_CONTENT "${ARGN}")
+    string(REPLACE ";" "" GENERATED_FILE_HEADER "${GENERATED_FILE_HEADER}")
+    string(REPLACE ";" "" GENERATED_FILE_TRAILER "${GENERATED_FILE_TRAILER}")
+    configure_file(
+        "${CFE_SOURCE_DIR}/cmake/cfe_generated_file.h.in"
+        "${FILE_NAME}"
+        @ONLY)
+
+endfunction(generate_c_headerfile)
+
+##################################################################
+#
+# FUNCTION: generate_config_includefile
+#
+# A function to create safe include file wrappers
+#
+# Rather than symlinking to the include file (which might not work the same on all platforms)
+# we can create a build-specific include file that just #include's the real file
+#
+# This also supports "stacking" multiple component files together by specifying more than one
+# source file for the wrapper.
+#
+# This function now accepts named parameters:
+#   FILE_NAME - the name of the file to write
+#   FALLBACK_FILE - if no files are found in "defs" using the name match, this file will be used instead.
+#   MATCH_SUFFIX - the suffix to match in the "defs" directory (optional)
+#   PREFIXES - a list of prefixes to match in the "defs" directory (optional)
```

Comment:
Should documentation  specify optional argument OUTPUT_DIRECTORY and its default location?

Reply:
See https://github.com/nasa/cFE/issues/1306